spec/integration/rails_spec.rb

```diff
@@ -118,211 +104,231 @@ def ping(recipient, subject)
         end
       end
 
-      @mock_intake = MockIntake.stub!
+      class TestAdapter < ElasticAPM::Transport::Connection
+        def write(payload)
+          EventCollector.catalog(JSON.parse(@metadata))
+          EventCollector.catalog JSON.parse(payload)
+        end
+      end
+
+      ElasticAPM::Transport::Worker.adapter = TestAdapter
```

Comment:
See above comment about including in `EventCollector` itself. I'm also not a fan of subclassing `Connection`. It feels like it's too easy to result in mysterious, unwanted behaviour. I'd much rather have exceptions from calling methods, `TestAdapter` doesn't implement (and then add them as noops where necessary.)

Reply:
I'm not sure I agree. Essentially what I'm doing here is using the real `Connection` class and mocking the `#write` method. Ideally, I could access the real `connection` object associated with a `worker` and mock the method there, but the current architecture doesn't allow me to do that; you can't access workers via `agent.transport`. (This is why I opened this [PR](https://github.com/elastic/apm-agent-ruby/pull/617), so that the workers owned their threads and were accessible via `agent.transport` and then could be mocked)

If we rewrite the methods that are necessary for this test to run, then those mock method definitions could eventually diverge from the real definitions, also resulting in mysterious behavior. For example, if I were to remove the inheritance, I would have to define an `initialize` method that matches the initialize method signature currently on `Connection`. I rely on the definition of `@metadata`, contained in `Connection#initialize`, so the mock `initialize` method definition can't be a noop. I'd rather use the real `Connection#initialize` method so that I can be sure that when I check the `@metadata` later, it was defined through the real code, not my mock object's definition.

Let me know what you think and if I've missed some points!

Reply:
With Sorbet types we could write an abstract interface and make sure any adapter would have the same api. I guess that's for another day 😅

Subclassing is fine. You make some good points 👍

Reply:
Yes, that is true! I was also going to make the argument that this is similar to how static languages mock objects (using abstract interfaces) in tests but didn't want to go overboard.
Thank you for hearing me out = )